man/man1/colcrt.1
.\" @(#)colcrt.1 1.1 92/07/30 SMI; from UCB 4.1
.TH COLCRT 1 "9 September 1987"
.SH NAME
colcrt \- filter nroff output for a terminal lacking overstrike capability
.SH SYNOPSIS
.B colcrt
.RB [ " \- " ] 
.RB [ " \-2 " ] 
.RI [ " filename " ".\|.\|. ]"
.SH DESCRIPTION
.IX  colcrt  ""  "\fLcolcrt\fP \(em document previewer"
.IX  "document production"  "colcrt command"  ""  "\fLcolcrt\fP command"
.IX  "preview documents, colcrt command"  ""  "\fLcolcrt\fP command"
.IX  "nroff utilities" colcrt "" "\fLcolcrt\fP \(em filter nroff output for CRT"
.LP
.B colcrt
provides virtual half and reverse
.SM LINEFEED
sequences for terminals without such capability, and on which overstriking
is destructive.  Half
.SM LINEFEED
characters and underlining
(changed to dashing
.RB ` \- ')
are placed on new lines in between the normal output lines.
.SH OPTIONS
.TP
.B \-
Suppress all underlining \(em especially useful for previewing
.I allboxed
tables from
.BR tbl (1).
.TP
.B \-2
Print all half
.SM LINEFEED
characters, effectively double spacing
the output.  Normally, a
minimal space output format is used which suppresses empty lines.
.B colcrt
never suppresses two consecutive empty lines, however.  The
.B \-2
option is useful for sending output to the line printer when the
output contains superscripts and subscripts
which would otherwise be invisible.
.SH EXAMPLE
.LP
A typical use of
.B colcrt
would be
.IP
.ft B
example% tbl exum2.n | nroff \-ms | colcrt \- | more
.ft R
.SH "SEE ALSO"
.BR col (1V),
.BR more (1),
.BR nroff (1),
.BR tbl (1),
.BR troff (1),
.BR ul (1)
.SH BUGS
.\".LP
.\"Should fold underlines onto
.\".SM SPACE
.\"characters even with the
.\".RB ` \- '
.\"option so that
.\"a true underline character would show; if we did this, however,
.\".B colcrt
.\"would not get rid of
.\".IR cu 'd
.\"underlining completely.
.LP
Cannot back up more than 102 lines.
.LP
General overstriking is lost;
as a special case
.RB ` \||\| '
overstruck with
.RB ` \|\-\| '
or underline becomes
.RB ` \|+\| '.
.LP
Lines are trimmed to 132 characters.
.LP
Some provision should be made for processing
superscripts and subscripts
in documents which are already double-spaced.


ucb/colcrt.c
/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1980 Regents of the University of California.\n\
 All rights reserved.\n";
#endif not lint

#ifndef lint
static	char sccsid[] = "@(#)colcrt.c 1.1 92/07/30 SMI"; /* from UCB 5.1 5/31/85 */
#endif not lint

#include <stdio.h>
/*
 * colcrt - replaces col for crts with new nroff esp. when using tbl.
 * Bill Joy UCB July 14, 1977
 *
 * This filter uses a screen buffer, 267 half-lines by 132 columns.
 * It interprets the up and down sequences generated by the new
 * nroff when used with tbl and by \u \d and \r.
 * General overstriking doesn't work correctly.
 * Underlining is split onto multiple lines, etc.
 *
 * Option - suppresses all underlining.
 * Option -2 forces printing of all half lines.
 */

char	page[267][132];

int	outline = 1;
int	outcol;

char	suppresul;
char	printall;

char	*progname;
FILE	*f;

main(argc, argv)
	int argc;
	char *argv[];
{
	register c;
	register char *cp, *dp;

	argc--;
	progname = *argv++;
	while (argc > 0 && argv[0][0] == '-') {
		switch (argv[0][1]) {
			case 0:
				suppresul = 1;
				break;
			case '2':
				printall = 1;
				break;
			default:
				printf("usage: %s [ - ] [ -2 ] [ file ... ]\n", progname);
				fflush(stdout);
				exit(1);
		}
		argc--;
		argv++;
	}
	do {
		if (argc > 0) {
			if ((f = freopen(argv[0], "r", stdin)) == NULL) {
				fflush(stdout);
				perror(argv[0]);
				exit (1);
			}
			argc--;
			argv++;
		}
		for (;;) {
			/*
			** ASSERT: outline < 266
			** (We must have this, since if c=='_', the line
			** page[outline+1] is written and page[266] is the
			** last allocated line in the array.)
			*/

			c = getc(stdin);
			if (c == -1) {
				pflush(outline);
				fflush(stdout);
				break;
			}
			switch (c) {
				case '\n':
					if (outline >= 264)
						pflush(62);
					outline += 2;
					outcol = 0;
					continue;
				case '\016':
					case '\017':
					continue;
				case 033:
					c = getc(stdin);
					switch (c) {
						case '9':
							if (outline >= 265)
								pflush(62);
							outline++;
							continue;
						case '8':
							if (outline >= 1)
								outline--;
							continue;
						case '7':
							outline -= 2;
							if (outline < 0)
								outline = 0;
							continue;
						default:
							continue;
					}
				case '\b':
					if (outcol)
						outcol--;
					continue;
				case '\t':
					outcol += 8;
					outcol &= ~7;
					outcol--;
					c = ' ';
				default:
					if (outcol >= 132) {
						outcol++;
						continue;
					}
					cp = &page[outline][outcol];
					outcol++;
					if (c == '_') {
						if (suppresul)
							continue;
						cp += 132;
						c = '-';
					}
					if (*cp == 0) {
						*cp = c;
						dp = cp - outcol;
						for (cp--; cp >= dp && *cp == 0; cp--)
							*cp = ' ';
					} else
						if (plus(c, *cp) || plus(*cp, c))
							*cp = '+';
						else if (*cp == ' ' || *cp == 0)
							*cp = c;
					continue;
			}
		}
	} while (argc > 0);
	fflush(stdout);
	exit(0);
	/* NOTREACHED */
}

plus(c, d)
	char c, d;
{

	return (c == '|' && d == '-' || d == '_');
}

int first;

pflush(ol)
	int ol;
{
	register int i, j;
	register char *cp;
	char lastomit;
	int l;

	l = ol;
	lastomit = 0;
	if (l > 266)
		l = 266;
	else
		l |= 1;
	for (i = first | 1; i < l; i++) {
		move(i, i - 1);
		move(i, i + 1);
	}
	for (i = first; i < l; i++) {
		cp = page[i];
		if (printall == 0 && lastomit == 0 && *cp == 0) {
			lastomit = 1;
			continue;
		}
		lastomit = 0;
		printf("%s\n", cp);
	}
	bcopy(page[ol], page, (267 - ol) * 132);
	bzero(page[267- ol], ol * 132);
	outline -= ol;
	outcol = 0;
	first = 1;
}

move(l, m)
	int l, m;
{
	register char *cp, *dp;

	for (cp = page[l], dp = page[m]; *cp; cp++, dp++) {
		switch (*cp) {
			case '|':
				if (*dp != ' ' && *dp != '|' && *dp != 0)
					return;
				break;
			case ' ':
				break;
			default:
				return;
		}
	}
	if (*cp == 0) {
		for (cp = page[l], dp = page[m]; *cp; cp++, dp++)
			if (*cp == '|')
				*dp = '|';
			else if (*dp == 0)
				*dp = ' ';
		page[l][0] = 0;
	}
}
